### DISCUSSION TITLE
State managment/data store

### DISCUSSION THREAD

**Discussion by zhuiks** [ezra-bible-app/ezra-bible-app#265] [Development]
It would be nice to have a global data store in place for easier state management  and unidirectional data flow.
![1-kPq4k51aqkrc6rmkYtHa4Q](https://user-images.githubusercontent.com/3584119/117488175-b6333680-af31-11eb-8670-8706edc89178.png)

Practically speaking what would be the best implementation of `options_menu.js`. 
Currently for each user's `<config-option>` click event `optionChange` is generated. There is a separate `eventHandler()` function for each option that does something (multiple things) in UI.

The code would be more clear if it was a unidirectional flow:

User click     -->     `<config-option>` optionChange action     -->     dispatcher/reducer updates the state    -->     new state     -->     UI updates based on state either via properties or listeners

Then the only thing UI components should worry about is reflecting the current state. And user events/actions handlers should only worry about updating the state. 

From my understanding that is common pattern for front-end now

Redux was one of the [suggestions](https://github.com/ezra-bible-app/ezra-bible-app/discussions/251#discussioncomment-700298)

**Comment by tobias-klein** [comment]
I like this proposal. Following standardized implementation patterns is a good thing! I'll look into this and we can discuss some more some time soon.

**Comment by zhuiks** [comment]
Just putting my thought as I'm thinking of it.

Based on my experience with React I've seen two approaches with the app state: 
1. **Centralized** state. Any component would have access to this global state and pull the props/data that are specific to the component.  That's [Redux way](https://redux.js.org/introduction/core-concepts)
2. **Context** state. One of the concepts of React is to have as little state exposed as needed. I.e. AppContainer component shouldn't know about selected verse in the VerseList component. Typically, certain part of the app state implemented via [ContextWrapper](https://reactjs.org/docs/context.html#when-to-use-context) component

Personally, I lean towards the second approach. I can see the benefits of having different layers of responsibility with LocaleContext (LocaleController), NightModeContext (ThemeController), BibleBrowserOptionsContext. But I want to read more about approaches for state management in other popular frameworks (Vue, Svetle)

**Reply by tobias-klein** [reply]
An additional state mgmt library to consider is [Zustand](https://zustand.surge.sh/).

**Reply by zhuiks** [reply]
Zustand seems to be nice. It's React oriented but not connected bounded by it

**Comment by zhuiks** [comment]
I've been thinking more through this. It seems like popular big guys ([Redux](https://redux.js.org/introduction/core-concepts), [MobX](https://mobx.js.org/observable-state.html)) that derived from [Flux](https://facebook.github.io/flux/docs/in-depth-overview#structure-and-data-flow) pattern might add unnecessary complexity to the frontend.
And then there are concepts of *reactivity* and *immutability*. Seems like people are talking about those when discussing state implementations.

I think the main idea is to have architecture that reflects the principle:
event ⇒ state update ⇒ view update as reaction to the state changes

- It's logical to update web components based view by setting up new attributes/properties
- events should be dispatched by component
- Having specific state/stores helps to decouple components. (i.e. BibleBrowser could be a separate component that can be used by other projects)

I found [this](https://github.com/fabiospampinato/store) little simple library...
